pkg/controllers/work/apply_controller.go

```diff
@@ -629,6 +672,144 @@ func buildManifestAppliedCondition(err error, action applyAction, observedGenera
+	var configMap v1.ConfigMap
+	annotations := obj.GetAnnotations()
+	configMapName, ok := annotations[configMapNameAnnotation]
+	if !ok {
+		return nil, errors.New("manifest doesn't have a config map")
+	}
+	if err := r.spokeClient.Get(ctx, types.NamespacedName{Name: configMapName, Namespace: fleetSystemNamespace}, &configMap); err != nil {
+		if apierrors.IsNotFound(err) {
+			// create config map since it's not found
+			manifestHash, err := computeManifestHash(obj)
+			if err != nil {
+				klog.ErrorS(err, "failed to compute manifest hash", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			lastModifiedConfig, err := computeModifiedConfiguration(obj)
+			if err != nil {
+				klog.ErrorS(err, "failed compute last applied config", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			if err := r.createConfigMap(ctx, configMapName, owner, manifestHash, lastModifiedConfig); err != nil {
+				klog.ErrorS(err, "cannot create config map for object", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+				return nil, err
+			}
+			// get newly created config map recursively, doesn't look clean
+			newConfigMap, err := r.getConfigMap(ctx, obj, owner)
+			if err != nil {
+				return nil, err
+			}
+			return newConfigMap, nil
+		} else {
+			klog.ErrorS(err, "failed to retrieve config map for object", "manifest", obj.GetName(), "config map", configMapName)
+			return nil, err
+		}
+	}
+	return &configMap, nil
+}
+
+func (r *ApplyWorkReconciler) updateConfigMap(ctx context.Context, configMap *v1.ConfigMap, obj *unstructured.Unstructured) error {
+	updatedManifestHash, err := computeManifestHash(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new manifest hash", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	updatedLastModifiedConfig, err := computeModifiedConfiguration(obj)
+	if err != nil {
+		// namespace for manifest can be empty
+		klog.ErrorS(err, "failed to compute new last modified configuration", "manifest", obj.GetName(), "namespace", obj.GetNamespace())
+		return err
+	}
+	configMap.Data[manifestHashKey] = updatedManifestHash
+	configMap.Data[lastAppliedConfigKey] = updatedLastModifiedConfig
+	if err := r.spokeClient.Update(ctx, configMap); err != nil {
+		klog.ErrorS(err, "failed to update config map", "configMap", configMap.Name)
+		return err
+	}
+	return nil
+}
+
+// We can calculate the manifestHash & lastAppliedConfig again wiping out previous state or make sure the annotations in the last applied config doesn't cause issue in patch.
+func (r *ApplyWorkReconciler) migrateToConfigMap(ctx context.Context, gvr schema.GroupVersionResource, obj *unstructured.Unstructured, owner metav1.OwnerReference) error {
+	// we can also use obj.getAnnotations().
+	annots, err := metadataAccessor.Annotations(obj)
```

Comment:
Hi Arvind! We might need to double-confirm this; for unstructured objects, this function seems to return a copy of the annotations map, but in this piece of code all the updates are in-place.

See also:
https://github.com/kubernetes/apimachinery/blob/4fbe8e4db39b17a53c0756caf58466baad871871/pkg/apis/meta/v1/unstructured/unstructured.go#L407
https://github.com/kubernetes/apimachinery/blob/4fbe8e4db39b17a53c0756caf58466baad871871/pkg/apis/meta/v1/unstructured/helpers.go#L166